src/gtk/overlay.cpp

```diff
@@ -0,0 +1,210 @@
+        // or the overlay widget is hidden.
+        return false;
+    }
+
+    if ( gtk_cairo_should_draw_window(cr, gtk_widget_get_window(widget)) )
+    {
+        cairo_surface_t* surface = overlay->GetCairoSurface();
+
+        if ( surface && cairo_surface_status(surface) == CAIRO_STATUS_SUCCESS )
+        {
+            cairo_set_source_surface(cr, surface, 0, 0);
+            cairo_paint(cr);
+        }
+    }
+
+    return false;
+}
+}
+
+// ----------------------------------------------------------------------------
+// wxOverlay
+// ----------------------------------------------------------------------------
+
+wxOverlayImpl::wxOverlayImpl()
+{
+    m_surface = NULL;
+    m_window = NULL;
+    m_cairoContext = NULL;
+    m_cairoSurface = NULL;
+}
+
+wxOverlayImpl::~wxOverlayImpl()
+{
+    if ( m_window )
+        m_window->Unbind(wxEVT_SIZE, &wxOverlayImpl::OnSize, this);
+
+    if ( m_cairoSurface )
+        cairo_surface_destroy(m_cairoSurface);
+
+    if ( m_surface )
+    {
+        g_signal_handlers_disconnect_by_data(m_surface, this);
+        gtk_widget_destroy(m_surface);
+        g_object_unref(m_surface);
+    }
+}
+
+bool wxOverlayImpl::IsOk() const
+{
+    return m_window != NULL;
+}
+
+void wxOverlayImpl::Init(wxDC *dc, int x, int y, int width, int height)
+{
+    wxCHECK_RET( dc, "NULL dc pointer" );
+    wxASSERT_MSG( !IsOk() , "You cannot Init an overlay twice" );
+
+    m_window = dc->GetWindow();
+    m_window->Bind(wxEVT_SIZE, &wxOverlayImpl::OnSize, this);
+
+    m_rect = wxRect(x, y, width, height);
+
+    int clientWidth, clientHeight;
+    m_window->GetClientSize(&clientWidth, &clientHeight);
+
+    if ( m_surface )
+    {
+        if ( gtk_widget_get_allocated_width(m_surface) != clientWidth ||
+             gtk_widget_get_allocated_height(m_surface) != clientHeight )
+        {
+            gtk_widget_set_size_request(m_surface, clientWidth, clientHeight);
+        }
+
+        gtk_widget_show(m_surface);
+        return;
+    }
+
+    // Create the surface:
+
+    wxCHECK_RET( m_window->GTKEnableOverlay(), "Overlay couldn't be enabled for m_window" );
+    GtkWidget* const overlay = m_window->m_widget;
+
+    m_surface = gtk_drawing_area_new();
+    gtk_widget_set_has_window(m_surface, FALSE);
+    gtk_widget_set_size_request(m_surface, clientWidth, clientHeight);
+    gtk_widget_set_halign(m_surface, GTK_ALIGN_FILL);
+    gtk_widget_set_valign(m_surface, GTK_ALIGN_FILL);
+    gtk_widget_set_vexpand(m_surface, TRUE);
+    gtk_widget_set_hexpand(m_surface, TRUE);
+    gtk_widget_set_can_focus(m_surface, FALSE);
+    gtk_widget_set_sensitive(m_surface, FALSE);
+
+    gtk_overlay_add_overlay( GTK_OVERLAY(overlay), m_surface );
+    gtk_overlay_set_overlay_pass_through( GTK_OVERLAY(overlay), m_surface, TRUE );
+
+    g_signal_connect(m_surface, "draw", G_CALLBACK(wxgtk_draw_overlay), this);
+
+    g_object_ref(m_surface);
+    gtk_widget_show(m_surface);
+
+    m_window->SetFocus();
+}
```

Comment:
@paulcor , @vadz 

Adding overlay to `m_window` is not supposed to alter the focus handling of that window in any way. but what i noticed is that:
- the first time i click on such window (`wxGrid` in fact, which supposed to set the focus on it) is not working at all!
- subsequent clicks work correctly.

I tried to trace the focus handling (with and without the `wxOverlay` being applied) but i'm still clueless what's going on here. any help/hint is much appreciated here of course.

For now, i'm adding an explicit call to `SetFocus()` which solves the problem. and if i am not mistaken, this can be an acceptable workaround knowing that `wxOverlay` expects `m_window` to be the focus window anyhow ?